core/src/main/kotlin/org/eln2/data/ComponentGraph.kt

```diff
@@ -0,0 +1,278 @@
+	fun vertexAdded(v: Vertex)
+	fun vertexRemoved(v: Vertex)
+}
+
+interface EdgeListener {
+	fun edgeAdded(e: Edge)
+	fun edgeRemoved(e: Edge)
+}
+
+open class ComponentGraph {
+	val vertices: MutableSet<Vertex> = mutableSetOf()
+	val edges: MutableSet<Edge> = mutableSetOf()
+	var components: MutableSet<ConnectedComponent> = mutableSetOf()
+	
+	val componentListeners: MutableSet<ComponentListener> = mutableSetOf()
+	val vertexListeners: MutableSet<VertexListener> = mutableSetOf()
+	val edgeListeners: MutableSet<EdgeListener> = mutableSetOf()
+	
+	fun addListener(listener: Any) {
+		if(listener is ComponentListener) componentListeners.add(listener)
+		if(listener is VertexListener) vertexListeners.add(listener)
+		if(listener is EdgeListener) edgeListeners.add(listener)
+	}
+	
+	fun removeListener(listener: Any) {
+		componentListeners.remove(listener)
+		vertexListeners.remove(listener)
+		edgeListeners.remove(listener)
+	}
+
+	internal open fun constructComponent(v: Vertex) = ConnectedComponent(v)
+	internal open fun constructVertex() = Vertex(this)
+	internal open fun constructEdge(a: Vertex, b: Vertex) = Edge(a, b)
+	
+	internal fun newVertex() = constructVertex().apply {
+		vertices.add(this)
+		components.add(component)
+		vertexListeners.forEach { it.vertexAdded(this) }
+	}
+	
+	internal fun removeVertex(v: Vertex) {
+		vertexListeners.forEach { it.vertexRemoved(v) }
+		disconnect(v)
+		vertices.remove(v)
+	}
+
+	internal fun connect(a: Vertex, b: Vertex): Edge {
+		assert(a.graph == this) { "CG.c: vertex a=$a not in graph $this" }
+		assert(b.graph == this) { "CG.c: vertex b=$b not in graph $this" }
+		
+		val edge = constructEdge(a, b)
+		edges.add(edge)
+		a.incident[b] = edge
+		b.incident[a] = edge
+
+		edgeListeners.forEach { it.edgeAdded(edge) }
+		
+		return edge
+	}
+	
+	internal fun disconnect(a: Vertex, b: Vertex): Boolean {
+		assert(a.graph == this) { "CG.d: vertex a=$a not in graph $this" }
+		assert(b.graph == this) { "Cg.d: vertex b=$a not in graph $this" }
+		
+		val edge = a.incident[b]
+		return if(edge != null) {
+			edgeListeners.forEach { it.edgeRemoved(edge) }
+			
+			a.incident.remove(b)
+			b.incident.remove(a)
+			edges.remove(edge)
+
+			true
+		} else {
+			false
+		}
+	}
+	
+	internal fun disconnect(v: Vertex) {
+		v.incident.keys.forEach {
+			disconnect(v, it)
+		}
+		assert(v.incident.isEmpty())
+	}
+
+	protected class Visitation(val vertex: Vertex, val phase: Phase, val root: Boolean = false) {
+		enum class Phase {
+			Down, Up
+		}
+	}
+
+	fun visitDepthFirst(
+		downPhase: (visited: Vertex, depth: Int, isRoot: Boolean, neighbors: MutableSet<Vertex>) -> Unit,
+		upPhase: (visited: Vertex, depth: Int) -> Unit,
+		start: Vertex? = null
+	) {
+		val unvisited = vertices.toMutableSet()  // XXX Relying on the implementation detail of a copy here...
+		var depth = 0
+		val stack = Stack<Visitation>()
+
+		while(!unvisited.isEmpty()) {
+			if(stack.isEmpty()) {
+				dprintln("CG.vDF: adding a new root")
```

Comment:
Use new debug stuff

Reply:
I'm looking at `master` for the "new debug stuff", and not seeing much in the way of substantial difference (mostly, it looks like it automatically extracts the calling class/method name). Is that what you're talking about?

Reply:
https://github.com/eln2/eln2/blob/master/core/src/main/kotlin/org/eln2/debug/DebugPrint.kt

TL;DR, remove the `CG.vDF` because we do that for you now.